File: JavaScript/src/graphClient.js
import { Client } from '@microsoft/microsoft-graph-client';
import { ClientSecretCredential } from '@azure/identity';

// Create a credential object using ClientSecretCredential
// This credential is used to authenticate requests to the Microsoft Graph API
const credential = new ClientSecretCredential(
  process.env.GRAPH_TENANT_ID, // Tenant ID of the Azure AD
  process.env.GRAPH_CLIENT_ID, // Client ID of the registered application
  process.env.GRAPH_CLIENT_SECRET // Client secret of the registered application
);

// Initialize the Microsoft Graph client with middleware
const graphClient = Client.initWithMiddleware({
  // Define the authentication provider
  authProvider: {
    // Function to fetch an access token for the Microsoft Graph API
    getAccessToken: async () => {
      // Request an access token for the Microsoft Graph API scope
      const token = await credential.getToken("https://graph.microsoft.com/.default");
      return token.token; // Return the token string
    },
  },
});

// Export the initialized Graph client for use in other modules
export default graphClient;

File: Python/graph_client.py
import os
from dotenv import load_dotenv
from azure.identity.aio import ClientSecretCredential
from msgraph import GraphServiceClient

# Load environment variables
load_dotenv()

def get_graph_client():
    """
    Initializes and returns a Microsoft Graph client using msgraph-sdk.
    """
    GRAPH_CLIENT_ID = os.getenv("GRAPH_CLIENT_ID")
    GRAPH_CLIENT_SECRET = os.getenv("GRAPH_CLIENT_SECRET")
    GRAPH_TENANT_ID = os.getenv("GRAPH_TENANT_ID")

    if not GRAPH_CLIENT_ID or not GRAPH_CLIENT_SECRET or not GRAPH_TENANT_ID:
        raise ValueError("Microsoft Graph credentials are missing in the environment variables.")

    # Authenticate using Azure Identity's ClientSecretCredential
    credential = ClientSecretCredential(
        tenant_id=GRAPH_TENANT_ID,
        client_id=GRAPH_CLIENT_ID,
        client_secret=GRAPH_CLIENT_SECRET
    )

    # Define the required scopes
    scopes = ['https://graph.microsoft.com/.default']

    # Initialize the Graph client
    graph_client = GraphServiceClient(credentials=credential, scopes=scopes)

    return graph_client